pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,360 @@
+		}
+		out = append(out, root)
+	}
+
+	return
+}
+
+// merge root's route with delegate's and the merged route number equals the delegate's.
+// if there is a conflict with root, the route is ignored
+func mergeHTTPRoutes(root *networking.HTTPRoute, delegate []*networking.HTTPRoute) []*networking.HTTPRoute {
+	root.Delegate = nil
+
+	out := make([]*networking.HTTPRoute, 0, len(delegate))
+	for _, subRoute := range delegate {
+		merged := mergeHTTPRoute(root, subRoute)
+		if merged != nil {
+			out = append(out, merged)
+		}
+	}
+	return out
+}
+
+// merge the two HTTPRoutes, if there is a conflict with root, the delegate route is ignored
+func mergeHTTPRoute(root *networking.HTTPRoute, delegate *networking.HTTPRoute) *networking.HTTPRoute {
+	// suppose there are N1 match conditions in root, N2 match conditions in delegate
+	// if match condition of N2 is a subset of anyone in N1, this is a valid matching conditions
+	merged, conflict := mergeHTTPMatchRequests(root.Match, delegate.Match)
+	if conflict {
+		log.Debugf("HTTPMatchRequests conflict: root route %s, delegate route %s", root.Name, delegate.Name)
+		return nil
+	}
+	delegate.Match = merged
+
+	if delegate.Name == "" {
+		delegate.Name = root.Name
+	}
+	if delegate.Rewrite == nil {
+		delegate.Rewrite = root.Rewrite
+	}
+	if delegate.Timeout == nil {
+		delegate.Timeout = root.Timeout
+	}
+	if delegate.Retries == nil {
+		delegate.Retries = root.Retries
+	}
+	if delegate.Fault == nil {
+		delegate.Fault = root.Fault
+	}
+	if delegate.Mirror == nil {
+		delegate.Mirror = root.Mirror
+	}
+	if delegate.MirrorPercent == nil {
+		delegate.MirrorPercent = root.MirrorPercent
+	}
+	if delegate.MirrorPercentage == nil {
+		delegate.MirrorPercentage = root.MirrorPercentage
+	}
+	if delegate.CorsPolicy == nil {
+		delegate.CorsPolicy = root.CorsPolicy
+	}
+	if delegate.Headers == nil {
+		delegate.Headers = root.Headers
+	}
+	return delegate
+}
+
+// return merged match conditions if not conflicts
+func mergeHTTPMatchRequests(root, delegate []*networking.HTTPMatchRequest) (out []*networking.HTTPMatchRequest, conflict bool) {
+	if len(root) == 0 {
+		return delegate, false
+	}
+
+	if len(delegate) == 0 {
+		return root, false
+	}
+
+	// each HTTPMatchRequest of delegate must find a superset in root.
+	// otherwise it conflicts
+	for _, subMatch := range delegate {
+		foundMatch := false
+		for _, rootMatch := range root {
+			if hasConflict(rootMatch, subMatch) {
+				log.Debugf("HTTPMatchRequests conflict: root %v, delegate %v", rootMatch, subMatch)
+				continue
+			}
+			// merge HTTPMatchRequest
+			out = append(out, mergeHTTPMatchRequest(rootMatch, subMatch))
+			foundMatch = true
+		}
+		if !foundMatch {
+			return nil, true
+		}
+	}
+	if len(out) == 0 {
+		conflict = true
+	}
+	return
+}
+
+func mergeHTTPMatchRequest(root, delegate *networking.HTTPMatchRequest) *networking.HTTPMatchRequest {
+	out := *delegate
+	if out.Name == "" {
+		out.Name = root.Name
```

Comment:
Nit: rather than using the delegate's name as is, do you think it makes sense to have root.Name-delegate.name ? This is because different delegates may have the same name. Envoy will complain if different routes have same name.

Reply:
or probably reject rds

Reply:
make sense